### PR TITLE
Remove all remaining unwraps

### LINKED ISSUE #428: [Bug] Potential panic when unwrapping in SignedStruct impl of KeyPackage
**Describe the bug**
The result of a `from_payload` call is unwrapped even though it could potentially yield an encoding error.

**Expected behavior**
Either the unwrap should be explained (i.e. why it's ok to unwrap here) or the error case should be handled.

**Actual behavior**
The return value is simply unwrapped without associated comment.

### LINKED ISSUE #438: Remove all unwraps
There are still a lot of unwraps in the code. This issue tracks removing all of them.

cc #261

### LINKED ISSUE #591: Refactor ciphersuite to remove unwraps
A few `unwrap`s are left in the ciphersuite module that need to be removed by refactoring the ciphersuite.

### PR DESCRIPTION
**Pull request by raphaelrobert**
Fixes #428 
Fixes #438 
Fixes #591

### CODE REVIEW

File: openmls/src/ciphersuite/mod.rs [hunk 1]
```diff
@@ -104,21 +104,26 @@ impl Ciphersuite {
             return Err(ConfigError::UnsupportedCiphersuite);
         }
 
-        Ok(Ciphersuite {
+        Ok(Self::new_from_supported(name))
+    }
+
+    /// This creates a Ciphersuite from one of the supported ciphersuite names.
+    /// This should only be used if it is clear the ciphersuite is used.
```
Comment (kkohbrok) [inline openmls/src/ciphersuite/mod.rs:111]: I'm having trouble understanding the second line. Should it be "... the ciphersuite is supported."?

File: openmls/src/key_packages/mod.rs [hunk 2]
```diff
@@ -108,31 +108,25 @@ impl KeyPackagePayload {
 pub struct KeyPackage {
     payload: KeyPackagePayload,
     signature: Signature,
-    encoded: Vec<u8>,
```
Comment (kkohbrok) [inline openmls/src/key_packages/mod.rs:111]: @franziskuskiefer I think you introduced that at some point to reduce serialization operations. Can you remember how big the impact was?

Also: Should we maybe do this in a separate commit/PR?

File: openmls/src/tree/secret_tree.rs [hunk 3]
```diff
@@ -168,21 +174,23 @@ impl SecretTree {
...
+            self.derive_down(ciphersuite, backend, n)?;
         }
         // Calculate node secret and initialize SenderRatchets
-        let node_secret = &self.nodes[index_in_tree.as_usize()]
-            .as_ref()
-            .unwrap()
-            .secret;
+        let node_secret = match &self.nodes[index_in_tree.as_usize()] {
+            Some(node) => &node.secret,
+            // We just derive all necessary nodes so this should not happen
```
Comment (kkohbrok) [inline openmls/src/tree/secret_tree.rs:182]:
```suggestion
            // We just derived all necessary nodes so this should not happen
```

**Review by kkohbrok — CHANGES_REQUESTED** [review]
Thanks! Good to finally be unwrap-free 🙂. Not sure about removing the KeyPackage encoding optimization, though.

File: openmls/src/key_packages/mod.rs (hunk 2 above)
Comment (raphaelrobert) [reply openmls/src/key_packages/mod.rs:111]: I ran benchmarks on both versions and there is no measurable difference (i.e. the tests are actually a tad faster after this change).

File: openmls/src/ciphersuite/mod.rs (hunk 1 above)
Comment (kkohbrok) [reply openmls/src/ciphersuite/mod.rs]: I see. But shouldn't this throw an error then, if the ciphersuite is not supported? Which makes me wonder what currently happens if this is called with an unsupported ciphersuite.

Taking a closer look, the issue of ciphersuite support seems a bit inconsistent. There are two functions `supported_ciphersuites`, which accesses the `Config` and `supported_ciphersuite_names`, which accesses a constant that is independent of the config.

So either way, I think this function should at least get a comment saying what happens if the input ciphersuite is not supported. Regarding the general inconsistency of the config, I'll file a follow-up.

**Review by kkohbrok — CHANGES_REQUESTED** [review]
Looks good to me with the exception of the `new_from_supported` issue.

**Comment by raphaelrobert** [comment]
@kkohbrok Thanks for filing the issue! Any other changes you would like?

**Review by kkohbrok — COMMENTED** [review]
I posted a comment under `new_from_supported`, but it probably got lost since the code is now outdated. I'll repost it here:

> I see. But shouldn't this throw an error then, if the ciphersuite is not supported? Which makes me wonder what currently happens if this is called with an unsupported ciphersuite.
> 
> Taking a closer look, the issue of ciphersuite support seems a bit inconsistent. There are two functions `supported_ciphersuites`, which accesses the `Config` and `supported_ciphersuite_names`, which accesses a constant that is independent of the config.
> 
> So either way, I think this function should at least get a comment saying what happens if the input ciphersuite is not supported. Regarding the general inconsistency of the config, I'll file a follow-up.

**Review by kkohbrok — APPROVED** [review]
Sounds good to me! Let's land it!

**Pull request merged**